### PR TITLE
Implement new makeCardPayment method to start support for 3DS2

### PR DESCRIPTION
**Pull request by manuel-huez**

### CODE REVIEW

File: src/processout/translator.ts [hunk 1]
```diff
@@ -39,6 +39,7 @@ module ProcessOut {
             "gateway.declined":        "The payment was declined.",
             "payment.declined":        "The payment was declined.",
             "payment.pending":         "The payment is currently pending, please wait a few minutes for it to fully go through.",
+            "three-d-s-2.fingerprint-timed-out": "The 3D-Secure fingerprinting timed out.",
```
Comment (louis-paul) [inline src/processout/translator.ts]:
```suggestion
            "three-d-s-2.fingerprint-timed-out": "The 3-D Secure fingerprinting timed out.",
```

File: src/processout/threeds.ts [hunk 2]
```diff
@@ -38,8 +38,8 @@ module ProcessOut {
                 throw new Exception("request.validation.error", "Please provide a source to be used to start the 3D-Secure flow.");
```
Comment (louis-paul) [inline src/processout/threeds.ts]:
```suggestion
                throw new Exception("request.validation.error", "Please provide a source to be used to start the 3-D Secure flow.");
```

File: src/processout/processout.ts [hunk 3]
```diff
@@ -363,6 +363,17 @@ module ProcessOut {
...
+            if (screen.colorDepth)
+                data.app_color_depth = Number(screen.colorDepth);
+            var language = navigator.language || (<any>navigator).userLanguage;
+            if (language)
+                data.app_language = language;
+            if (screen.height)
+                data.app_screen_height = screen.height;
+            if (screen.width)
+                data.app_screen_width = screen.width;
+            data.time_zone_offset = Number(new Date().getTimezoneOffset());
```
Comment (louis-paul) [inline src/processout/processout.ts]: Missing: whether Java is enabled

File: src/processout/processout.ts [hunk 4]
```diff
@@ -638,5 +690,77 @@ module ProcessOut {
...
+                    return;
+                }
+
+                var nextStep = function(data: any): void {
+                    options.gatewayRequestSource = data;
+                    this.makeCardPayment(invoiceID, cardID, options, success, error);
+                }.bind(this);
+
+                switch (data.customer_action.type) {
+                case "url":
```
Comment (louis-paul) [inline src/processout/processout.ts]: As discussed we will try to deprecate `url` and eventually merge it with `redirect`, right?

File: src/processout/actionhandler.ts [hunk 5]
```diff
@@ -49,12 +49,15 @@ module ProcessOut {
...
     }
 
-    export class ActionHandlerOptions {
-        // defines if we want to use an iframe or not
-        public useIFrame: boolean;
+    export enum ActionFlow {
+        NewTab,
+        NewWindow,
+        IFrame,
+        Fingerprint,
```
[the remark this answers is not in the document]
Comment (manuel-huez) [reply src/processout/actionhandler.ts]: The fingerprint has one specific behavior which is its timeout handling after 10s

File: src/processout/processout.ts (hunk 4 above)
Comment (manuel-huez) [reply src/processout/processout.ts]: Yep

File: src/processout/actionhandler.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (louis-paul) [reply src/processout/actionhandler.ts]: Maybe `FingerprintIframe`?

File: src/processout/actionhandler.ts [hunk 6]
```diff
@@ -64,42 +67,45 @@ module ProcessOut {
         // gatewayLogo is shown when the action is done on another tab or window
         public gatewayLogo?: string;
 
-        public static ThreeDSFlow = "three-d-s-flow";
+        public static ThreeDSChallengeFlow = "three-d-s-challenge-flow";
+        public static ThreeDSFingerprintFlow = "three-d-s-fingerprint-flow";
```
[the remark this answers is not in the document]
Comment (manuel-huez) [reply src/processout/actionhandler.ts]: We also expect gateway names to be sent, which are strings passed around by the API. Basically, the actionType can be 3DS but can also be for example paypal

[the remark this answers is not in the document]
Comment (louis-paul) [reply src/processout/actionhandler.ts]: Thanks for the answer!

File: src/processout/actionhandler.ts [hunk 7]
```diff
@@ -295,6 +289,91 @@ module ProcessOut {
...
+            var w = this.options.newWindowWidth;
+            var y = window.top.outerHeight / 2 + window.top.screenY - (h / 2);
+            var x = window.top.outerWidth / 2 + window.top.screenX - (w / 2);
+            return this.setupWindowObject(window.open(url, '',
+                `menubar=false,toolbar=false,width=${w},height=${h},top=${y},left=${x}`));
+        }
+
+        protected setupWindowObject(newWindow: any): any {
+            if (!newWindow) {
+                return {};
```
[the remark this answers is not in the document]
Comment (manuel-huez) [reply src/processout/actionhandler.ts]: We didn't lose it, it got transferred higher in the stack: https://github.com/processout/processout.js/pull/42/files#diff-daa1f80c83b45759b65fdd5fc50a9cc6R253

[the remark this answers is not in the document]
Comment (louis-paul) [reply src/processout/actionhandler.ts]: Cool!

File: src/processout/actionhandler.ts [hunk 8]
```diff
@@ -295,6 +289,91 @@ module ProcessOut {
...
+                    success(data);
+                }, function(err: Exception): void {
+                    if (timeout) clearTimeout(timeout);
+                    error(err);
+                });
+            return this;
+        }
+
+        protected createNewWindow(url: string): any {
+            return this.setupWindowObject(window.open(url, '_blank'));
```
[the remark this answers is not in the document]
Comment (manuel-huez) [reply src/processout/actionhandler.ts]: ah yeah, good point. thanks!

**Pull request merged**